### DISCUSSION TITLE
Help needed with making a tutorial.

### DISCUSSION THREAD

**Discussion by BakungaBronson** [reach-sh/reach-lang#1026] [Q&A] [answered]
Hello,
I recently finished up my Reach app and I would like to make a tutorial for the application. How can I get started and how will I eventually be able to submit it for review?

**Comment by jeapostrophe** [comment, marked as answer]
@TheChronicMonster Please reply

**Reply by TheChronicMonster** [reply]
Hello, thank you for this question. 

Allow me to begin with a little background on the types of curricula that can be created to teach others how to build your DApp. 

## Types of Curricula

Three of the primary types are guides, workshops, and tutorials. I'll explain what I mean for each of these to help you gain a sense of which type you'd like to build. 

**Guides**

A guide is a problem or idea-centered document that focuses on a problem the user may want to solve. For example, "No one finished a game of chess with me so I created this contract that ensures players play till the end of a match or lose their staked coins, and this is how you can create your own contract. 

A guide tends to be the shortest of these three types of curricula. It's not expected to contain exhaustive explanations of every line of code, but it does provide links to the codebase and discusses concepts used to create the application. 

**Workshops**

A workshop is project-oriented and helps the reader develop their thought process while also building the codebase. A workshop tends to be longer than a guide, but isn't as exhaustive in explanations as a tutorial. An example of a good workshop is [Hash Lock](https://docs.reach.sh/workshop/hash-lock/#workshop-hash-lock). This document helps the reader think about how they'll build their DApp. It encourages the user to do the heavy lifting and then supplies the author's answers so the reader can compare their notes.

**Tutorials**

A tutorial is a foundation for independent work. Additionally, a tutorial, such as [Rock, Paper, Scissors](https://docs.reach.sh/tut/rps/#tut), is an exhaustive document that holds the reader's hand through the process from start to finish. While every tutorial does not need to be as extensive as RPS, the document should be clear in its assumptions and scope. 

**Content**

For example, we could reimagine RPS as a tutorial that assumes the reader has already installed Reach and has already been building command-line test suites. In this case, we might limit our scope to integrating a working backend into a responsive React frontend. Alternatively, we might assume this is the first tutorial everyone will use with reach, but limit its scope as a quick startup guide. In this instance, we might introduce Reach, explain how to install the program, and then produce one or two quick working examples and conclude with a link to the next tutorial we think the reader should work on. 

No matter the assumptions or scope, the tutorial is quite detailed in explaining each line of code, and should also attempt to explain core concepts along the way. So a good tutorial not only teaches a student how to build a program, it also teaches the student how to think about the program they are creating. 

A tutorial with a large scope will feature multiple steps with directories containing code samples that reflect each of those steps. For example, in RPS, we begin with a simple scope of, "This is how you build a prototype Reach DApp," and then we explain components as we add them in each consecutive step. It's almost like a set of LEGO instructions. 

A good tutorial adds building blocks one after the other to expand the reader's understanding of the program and how they might build their own programs in the future. 
 
Regardless of format, the user should have access to working source code by the end of the guide.

## Curricula Format

To answer the question that people interested in writing their first tutorial really want to know: "How can I write a tutorial?" what you're really asking is, "How can I write a tutorial that provides so much value that people want to work through it till the end."

**Value Statement (The Why)**
Start with Why. In everything that requires persuasion, state why the information is valuable to the reader. Explain why the reader should be interested and why this is important. Essentially, you're making a value statement directly to the reader.

**Scope**

Next, begin with the end in mind. Know the scope of the document. How far will you carry the reader? Set the reader's expectations at the start of the document. "In this tutorial, you will learn how to install Reach and build a command-line DApp with two `Participants`, 'Alice' and 'Bob' who play a simple game of Rock, Paper, Scissors." Or, "In this tutorial, you will use a working Reach DApp and convert the frontend to a responsive web app using React."

Stating the scope sets expectations so reader's aren't disappointed. 

**Assumptions**

Understand and state your assumptions. Then begin where your assumptions end. i.e. if you assume the user has already installed Reach, then say so, provide a link to [installation instructions](https://docs.reach.sh/quickstart/) and then begin the tutorial at the next step after installation. 

**Learning Objectives**

Tell the reader what they will learn. Using assumptions and scope as constraints, explicitly tell your reader what they will learn by the end of the tutorial. These are your learning objectives.

**Architecture**

An introduction that includes the value statement, scope, assumptions, and learning objectives sets your tutorial up for success. 

Then build the program. How detailed you are in your explanations depends on what type of learning document you're creating. The most important attributes here are that the writing is clear and as concise as possible, that there is thoughtful flow from one thought to the next and that access to solutions are provided at the end of each section. 

Explain new concepts as they are discovered. A new concept is any paradigm or utility that is outside the introductory assumptions. Natural learning opportunities are excellent places to introduce new concepts. 

An error message is a learning opportunity. A need to change the architecture of the program is a learning opportunity. Use learning opportunities to explain concepts that are critical to becoming a successful engineer. 

At the end of each section, tell the reader what they've just learned. Remind them of the learning objectives that have been covered and when applicable, point them to resources to learn more. 

Repeat that process for each section of the document. 

At the conclusion, review the entire document and remind the reader of all of the learning objectives that they've accomplished. Then, tell them what a next step is. Often this includes adding to the existing codebase, building their own project, and/or moving on to a new learning document. 

(This is the tl;dr) 
In summary, the order of building a learning document is:

* Introduction
  * Why this is important (Value statement)
  * Scope
  * Assumptions
  * Learning Objectives
* Body
  * Sections
    * Introduce section
    * Build the program
    * Explain atomic steps
    * Demystify new concepts as they are introduced
    * Repeat steps 2-4 until the end of section is reached
    * Conclude section with code samples and explanation of completed learning objectives
  * Repeat steps in "Sections" block until scope is satisfied
* Conclusion
  * Tell the reader what they've learned
  * Provide a summary of the codebase
  * Share next steps
    * Code base
    * Documentation
    * Learning resources
    * What's next

## Final Things

**Word Structure & Tools**

Perfect spelling, proper grammar and word flow is critically important. The document should be as smooth to read as possible. An editor to act as proofreader can be an enormous asset. Free tools such as [Grammarly](https://www.grammarly.com/) and VSCode's [Spell Checker Extension](https://marketplace.visualstudio.com/items?itemName=streetsidesoftware.code-spell-checker) are helpful, as well. The specific tool or brand isn't as important as finding one that works for you. 

**Where to Submit**

We encourage Reach developers to publish their guides, workshops, and tutorials in public spaces on the internet. Currently, two popular sites include [Medium](https://medium.com) and [Hashnode](https://hashnode.com).
 
Feel free to ask your colleagues and member's of Reach's [Discord](https://discord.gg/jXyTScWA) community to help with proofreading and making suggestions for improvement. 

**Promotions**

To help promote your work, Marketing and communications at Reach is pleased to broadcast your polished tutorials on our social channels. Let us know in our Discord's #General channel when you're ready to publish so we can help share your content!
